src/charm.py

```diff
@@ -746,11 +752,21 @@ def _build_replication(self, primary: bool) -> Layer:
 
     @property
     def grafana_version(self):
-        """Grafana server version."""
-        info = self.grafana_service.build_info
-        if info:
-            return info.get("version", None)
-        return None
+        """Grafana server version.
+
+        Returns:
+            A string equal to the Grafana server version.
+        """
+        container = self.containers["workload"]
+        if not container.can_connect():
+            return None
+        version_output, _ = container.exec(["grafana-server", "-v"]).wait_output()
+        # Output looks like this:
+        # Version 8.2.6 (commit: d2cccfe, branch: HEAD)
+        result = re.search(r"Version (\d*\.\d*\.\d*)", version_output)
+        if result is None:
+            return result
+        return result.group(1)
```

Comment:
This is typically written with a walrus:
```suggestion
        if match := re.search(r"Version (\d*\.\d*\.\d*)", version_output):
            return match.group(1)
        return None
```

Reply:
As said many times before, we should avoid this if possible. That would be literally the only occurrence of a walrus inside the Grafana codebase.

It costs us very little and potentially benefits quite a bit (similar to the request for a machine version of the `grafana-agent` charm, which we cannot guarantee will run on a recent version of Python) if a scenario comes up where the charm codebase also needs to run with older versions.

Reply:
Yeah I was avoiding based on @rbarry82's argument.